File: scripts/fetch_wowhead_icons.py
#!/usr/bin/env python3
"""
Obtiene el nombre del icono correcto para cada material desde Wowhead (TBC item XML)
y actualiza app/lib/alchemy-tbc-materials.ts con los iconos únicos por item.

Uso: python scripts/fetch_wowhead_icons.py

Requisito: pip install requests
"""

import re
import time
from pathlib import Path

import requests

ROOT = Path(__file__).resolve().parent.parent
MATERIALS_TS = ROOT / "app" / "lib" / "alchemy-tbc-materials.ts"
WOWHEAD_XML = "https://www.wowhead.com/tbc/item={}&xml=1"


def get_icon_for_item(item_id: int) -> str | None:
    """Obtiene el nombre del icono del item desde Wowhead XML."""
    url = WOWHEAD_XML.format(item_id)
    headers = {"User-Agent": "Mozilla/5.0 (compatible; profesiones-wow/1.0)"}
    try:
        r = requests.get(url, headers=headers, timeout=10)
        r.raise_for_status()
        # <icon displayId="0">inv_misc_herb_10</icon>
        m = re.search(r"<icon[^>]*>([^<]+)</icon>", r.text)
        if m:
            return m.group(1).strip().lower().replace("\\", "")
    except Exception:
        pass
    return None


def main() -> None:
    content = MATERIALS_TS.read_text(encoding="utf-8")
    # Encontrar cada material: key: { name: "...", itemId: N, icon: "..." }
    pattern = re.compile(
        r"(\w+):\s*\{\s*name:\s*\"([^\"]+)\",\s*itemId:\s*(\d+),\s*icon:\s*\"([^\"]+)\"",
        re.MULTILINE,
    )
    updates = []
    for m in pattern.finditer(content):
        key, item_id, old_icon = m.group(1), int(m.group(3)), m.group(4)
        new_icon = get_icon_for_item(item_id)
        if new_icon:
            updates.append((m.start(), m.end(), m.group(0), key, old_icon, new_icon))
        else:
            print(f"  (sin icono) {key} itemId={item_id}")
        time.sleep(0.35)
    if not updates:
        print("No se pudo obtener ningún icono de Wowhead (revisa conexión).")
        return
    new_content = content
    for start, end, old_block, key, old_icon, new_icon in reversed(updates):
        new_block = old_block.replace(f'icon: "{old_icon}"', f'icon: "{new_icon}"')
        new_content = new_content[:start] + new_block + new_content[end:]
        if new_icon != old_icon:
            print(f"  {key}: {old_icon} -> {new_icon}")
    MATERIALS_TS.write_text(new_content, encoding="utf-8")
    print(f"\nActualizados {len(updates)} iconos en {MATERIALS_TS.name}. Vuelve a ejecutar download_wow_icons.py para descargar las imágenes.")


if __name__ == "__main__":
    main()

File: scripts/fetch-wowhead-icons.mjs
#!/usr/bin/env node
/**
 * Obtiene el nombre del icono de cada ítem desde Wowhead (tbc/item=ID?xml)
 * y actualiza app/lib/alchemy-tbc-materials.ts con el icon correcto.
 * Ejecutar: node scripts/fetch-wowhead-icons.mjs
 */

import { readFileSync, writeFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const MATERIALS_PATH = join(__dirname, "..", "app", "lib", "alchemy-tbc-materials.ts");

const ITEM_IDS_BY_KEY = {
  peacebloom: 2447,
  silverleaf: 2449,
  emptyVial: 3371,
  briarthorn: 2450,
  bruiseweed: 2453,
  leadedVial: 3372,
  mageroyal: 785,
  stranglekelp: 3820,
  liferoot: 3357,
  kingsblood: 3356,
  goldthorn: 3821,
  wildSteelbloom: 3355,
  sungrass: 8838,
  khadgarWhisker: 3358,
  crystalVial: 17194,
  arthasTears: 8839,
  blindweed: 8831,
  goldenSansam: 13464,
  mountainSilversage: 13463,
  imbuedVial: 18256,
  felweed: 22785,
  dreamingGlory: 22786,
  netherbloom: 22787,
  nightmareVine: 22792,
  terocone: 22789,
  dreamfoil: 13462,
  ancientLichen: 22790,
  icecap: 13467,
  elementalFire: 7068,
  dreamDust: 11176,
  ragveil: 22791,
  firefinSnapper: 6359,
};

function extractIconFromXml(xml) {
  const m = xml.match(/<icon[^>]*>([^<]+)<\/icon>/i);
  if (!m) return null;
  return m[1].replace(/\\/g, "").trim().toLowerCase();
}

async function fetchIconForItem(itemId) {
  const url = `https://www.wowhead.com/tbc/item=${itemId}?xml`;
  try {
    const res = await fetch(url);
    const xml = await res.text();
    return extractIconFromXml(xml);
  } catch (e) {
    console.warn(`Error fetching item ${itemId}:`, e.message);
    return null;
  }
}

async function main() {
  const results = {};
  const keys = Object.keys(ITEM_IDS_BY_KEY);
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    const itemId = ITEM_IDS_BY_KEY[key];
    process.stderr.write(`[${i + 1}/${keys.length}] ${key} (${itemId})... `);
    const icon = await fetchIconForItem(itemId);
    results[key] = icon;
    process.stderr.write(icon ? `${icon}\n` : "FAIL\n");
    await new Promise((r) => setTimeout(r, 300));
  }

  const content = readFileSync(MATERIALS_PATH, "utf8");
  let newContent = content;
  for (const [key, icon] of Object.entries(results)) {
    if (!icon) continue;
    const escapedKey = key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const re = new RegExp(
      `(${escapedKey}:\\s*\\{\\s*name:\\s*"[^"]+",\\s*itemId:\\s*\\d+,\\s*icon:\\s*)"[^"]*"`,
      "g"
    );
    newContent = newContent.replace(re, `$1"${icon}"`);
  }
  writeFileSync(MATERIALS_PATH, newContent);
  console.log("Actualizado", MATERIALS_PATH);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
